sockeye/init_embedding.py

```diff
@@ -0,0 +1,108 @@
+            -e embed-in-src.npy embed-in-tgt.npy    \
+            -i vocab-in-src.json vocab-in-tgt.json   \
+            -o vocab-out-src.json vocab-out-tgt.json  \
+            -n source_embed_weight target_embed_weight \
+            -f params.init
+
+Optional arguments:
+
+    --embeddings, -e
+        list of input embedding weights in .npy format
+        shape=(vocab-in-size, embedding-size)
+
+    --vocabularies-in, -i
+        list of input vocabularies as token-index dictionaries in .json format
+
+    --vocabularies-out, -o
+        list of output vocabularies as token-index dictionaries in .json format
+        They can be generated using sockeye.vocab before actual Sockeye training.
+
+    --names, -n
+        list of Sockeye parameter names for embedding weights
+        Most common ones are source_embed_weight, target_embed_weight and source_target_embed_weight.
+
+    Sizes of above 4 lists should be exactly the same - they are vertically aligned.
+
+    --file, -f
+        file to write initialized parameters
+
+    --encoding, -c
+        open input vocabularies with specified encoding (default: utf-8)
+"""
+
+import argparse
+import sys
+
+import numpy as np
+import mxnet as mx
+
+from sockeye.log import setup_main_logger, log_sockeye_version
+from . import arguments
+from . import utils
+from . import vocab
+
+logger = setup_main_logger(__name__, console=True, file_logging=False)
+
+def main():
+    """
+    Commandline interface to initialize Sockeye embedding weights with pretrained word representations.
+    """
+    log_sockeye_version(logger)
+    params = argparse.ArgumentParser(description='Quick usage: python3 -m contrib.utils.init_embedding '
+                                                 '-e embed-in-src.npy embed-in-tgt.npy '
+                                                 '-i vocab-in-src.json vocab-in-tgt.json '
+                                                 '-o vocab-out-src.json vocab-out-tgt.json '
+                                                 '-n source_embed_weight target_embed_weight '
+                                                 '-f params.init')
+    arguments.add_init_embedding_args(params)
+    args = params.parse_args()
+
+    if len(args.embeddings) != len(args.vocabularies_in) or \
+       len(args.embeddings) != len(args.vocabularies_out) or \
+       len(args.embeddings) != len(args.names):
+           logger.error("Exactly the same number of 'input embedding weights', 'input vocabularies', "
+                         "'output vocabularies' and 'Sockeye parameter names' should be provided.")
+           sys.exit(1)
+
+    params = {} # type: Dict[str, mx.nd.NDArray]
+    for embed_file, vocab_in_file, vocab_out_file, name in zip(args.embeddings, args.vocabularies_in, \
+                                                               args.vocabularies_out, args.names):
+        logger.info('Loading input embedding weight: %s', embed_file)
```

Comment:
It would make sense to refactor the logic of building an embedding matrix given two vocabularies, an existing embedding into a separate function in this file;
It could also take an mxnet initializer to initialize the output, instead of relying on numpy (and using a normal distribution for initialization). You can look at sockeye/initializers.py for that.

The use case could be to use the embedding initialization code directly in a sockeye.train call without having to pre-built the the vocabulary. This does not have to be part of this PR, but would make the code more reusable.
Additionally, its easier to write a test :)

Reply:
Could you please point me how to initialize an mx.nd.NDArray directly? Do I have to create a computation graph in order to use mxnet initializers?

Reply:
You can do the following:
```python
# assume an initializer like:
init = mx.init.Xavier(rnd_type="uniform")
# the array you want to initialize:
w = mx.nd.zeros((3,2))
w_desc = mx.init.InitDesc("embed_weight")
init(w_desc, w)
print(w)
```